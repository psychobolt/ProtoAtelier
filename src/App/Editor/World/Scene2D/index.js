export * from './Scene2D.component';
export { default } from './Scene2D.component';
