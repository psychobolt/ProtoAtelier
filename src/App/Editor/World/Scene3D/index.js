export { default } from './Scene3D.component';
