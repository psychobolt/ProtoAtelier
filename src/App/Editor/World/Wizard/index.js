export { default } from './Wizard.component';
