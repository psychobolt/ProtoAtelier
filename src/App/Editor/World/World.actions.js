import Hashids from 'hashids';

export const Actions = {
  UPDATE_GRAVITY: 'updateGravity',
  SET_PIXELS_PER_METER: 'setPixelsPerMeter',
  ADD_BODY: 'addBody',
  REMOVE_BODIES: 'removeBodies',
  UPDATE_BODIES: 'updateBodies',
  ADD_BODY_LINKS: 'addBodyLinks',
  ADD_FIXTURES: 'addFixtures',
  REMOVE_FIXTURES: 'removeFixtures',
  UPDATE_FIXTURES: 'updateFixtures',
  SELECT_BODY: 'selectBody',
  ADD_JOINT: 'addJoint',
  REMOVE_JOINTS: 'removeJoints',
  UPDATE_JOINTS: 'updateJoints',
  ADD_ANCHOR: 'addAnchor',
  REMOVE_ANCHORS: 'removeAnchors',
  UPDATE_ANCHORS: 'updateAnchors',
  TOGGLE_PAUSE: 'togglePause',
  TOGGLE_TRACKING: 'toggleTracking',
  UPDATE_MOUSE_FORCE: 'updateMouseForce',
  UPDATE_SPEED: 'updateSpeed',
  RESET: 'reset',
};

export const updateGravity = gravity => ({ type: Actions.UPDATE_GRAVITY, payload: gravity });

export const setPixelsPerMeter = value => ({
  type: Actions.SET_PIXELS_PER_METER, payload: { value },
});

export const addBody = skipHistory => ({
  type: Actions.ADD_BODY,
  payload: {
    id: new Hashids('body').encode(new Date().getTime()),
    fixtureIds: [],
    type: 'static',
    linearVelocity: { x: 0, y: 0 },
    linearDamping: 0,
    angularDamping: 0,
    bullet: false,
  },
  meta: { skipHistory },
});

export const removeBodies = ids => ({ type: Actions.REMOVE_BODIES, payload: { ids } });

export const updateBodies = bodies => ({ type: Actions.UPDATE_BODIES, payload: bodies });

export const addFixtures = (bodyId, fixtures, skipHistory) => ({
  type: Actions.ADD_FIXTURES,
  payload: {
    bodyId,
    fixtures: fixtures.map((fixture, index) => ({
      id: new Hashids('fixture').encode(index, new Date().getTime()),
      bodyId,
      density: 0,
      friction: 0,
      restitution: 0,
      ...fixture,
    })),
  },
  meta: { skipHistory },
});

export const removeFixtures = (ids, skipHistory) => ({
  type: Actions.REMOVE_FIXTURES, payload: { ids }, meta: { skipHistory },
});

export const updateFixtures = fixtures => ({ type: Actions.UPDATE_FIXTURES, payload: fixtures });

export const selectBody = id => ({
  type: Actions.SELECT_BODY, payload: { id }, meta: { skipHistory: true },
});

export const addJoint = options => ({
  type: Actions.ADD_JOINT,
  payload: {
    id: new Hashids('joint').encode(new Date().getTime()),
    anchors: {},
    enableMotor: false,
    maxMotorTorque: 0,
    motorSpeed: 0,
    ...options,
  },
});

export const removeJoints = ids => ({ type: Actions.REMOVE_JOINTS, payload: { ids } });

export const updateJoints = joints => ({ type: Actions.UPDATE_JOINTS, payload: joints });

export const addAnchor = (jointId, anchor) => ({
  type: Actions.ADD_ANCHOR,
  payload: {
    id: new Hashids('anchor').encode(new Date().getTime()),
    jointId,
    ...anchor,
  },
});

export const removeAnchors = (jointId, anchorIds) => ({
  type: Actions.REMOVE_ANCHORS, payload: { jointId, anchorIds },
});

export const updateAnchors = (jointId, anchors) => ({
  type: Actions.UPDATE_ANCHORS, payload: { jointId, anchors },
});

export const togglePause = () => ({ type: Actions.TOGGLE_PAUSE });

export const toggleTracking = () => ({ type: Actions.TOGGLE_TRACKING });

export const updateMouseForce = force => ({ type: Actions.UPDATE_MOUSE_FORCE, payload: force });

export const updateSpeed = speed => ({ type: Actions.UPDATE_SPEED, payload: speed });

export const reset = () => ({ type: Actions.RESET, payload: { id: new Hashids('world').encode(new Date().getTime()) } });
