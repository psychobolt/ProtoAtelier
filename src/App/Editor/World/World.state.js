import Hashids from 'hashids';

export default {
  id: new Hashids('world').encode(new Date().getTime()),
  paused: false,
  tracking: false,
};
