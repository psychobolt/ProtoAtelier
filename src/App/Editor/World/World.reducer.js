import initialState from './World.state';
import { Actions } from './World.actions';

export default (state = initialState, action) => {
  switch (action.type) {
    case Actions.TOGGLE_TRACKING: return { ...state, tracking: !state.tracking };
    case Actions.TOGGLE_PAUSE: return { ...state, paused: !state.paused };
    case Actions.RESET: return { ...state, id: action.payload.id };
    default: return state;
  }
};
