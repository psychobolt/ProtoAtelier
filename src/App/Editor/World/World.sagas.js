import { all, put, select, takeEvery } from 'redux-saga/effects';

import { getCanvas, getCanvasSettings, getWorld } from 'App/App.selectors';

import { Actions as CanvasActions, linkPath, linkLayer } from '../Canvas/Canvas.actions';
import { Actions as WorldActions, addBody, addFixtures, removeFixtures, selectBody } from './World.actions';

function* addFixture(bodyId, pathId) {
  const effect = addFixtures(bodyId, [{ pathId }], true);
  yield put(linkPath(pathId, { fixtureId: effect.payload.fixtures[0].id }));
  yield put(effect);
}

function* onAddLayer(action) {
  const settings = yield select(getCanvasSettings);
  if (settings.autoSync) {
    const { id } = action.payload;
    const effect = addBody(true);
    yield put(linkLayer(id, { bodyId: effect.payload.id }));
    yield put(effect);
  }
}

function* onAddPath(action) {
  const { autoSync, layerLinks } = yield select(getCanvas);
  const { id: pathId, layer, visible } = action.payload;
  const { bodyId } = layerLinks[layer];
  if (visible === false) {
    return;
  }
  if (autoSync) yield addFixture(bodyId, pathId);
}

function* onUpdatePaths(action) {
  const { autoSync, pathLinks, layerLinks } = yield select(getCanvas);
  if (autoSync) {
    yield all(action.payload.map(path => {
      const { id: pathId, layer, closed } = path;
      const pathLink = pathLinks[pathId] || {};
      if (pathLink.fixtureId || !closed) {
        return null;
      }
      const { bodyId } = layerLinks[layer];
      return addFixture(bodyId, pathId);
    }));
  }
}

function* onRemovePaths(action) {
  const { fixtureIds, fixtures } = yield select(getWorld);
  yield put(removeFixtures(fixtureIds.filter(id =>
    action.payload.ids.includes(fixtures[id].pathId)), true));
}

function* onAddBody(action) {
  yield put(selectBody(action.payload.id));
}

function* onRemoveLayers() {
  const { fixtures, fixtureIds } = yield select(getWorld);
  const { pathIds } = yield select(getCanvas);
  const ids = fixtureIds.filter(id => !pathIds.includes(fixtures[id].pathId));
  if (ids.length) {
    yield put(removeFixtures(ids, true));
  }
}

export default function* saga() {
  yield takeEvery(CanvasActions.ADD_LAYER, onAddLayer);
  yield takeEvery(CanvasActions.ADD_PATH, onAddPath);
  yield takeEvery(CanvasActions.UPDATE_PATHS, onUpdatePaths);
  yield takeEvery(CanvasActions.REMOVE_LAYERS, onRemoveLayers);
  yield takeEvery(CanvasActions.REMOVE_PATHS, onRemovePaths);
  yield takeEvery(WorldActions.ADD_BODY, onAddBody);
}
