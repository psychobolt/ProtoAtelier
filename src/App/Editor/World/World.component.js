// @flow
import React from 'react';
import { Switch, Route } from 'react-router-dom';

import withResizableContainer from 'Framework/ReactResizableContainer';
import Wizard from './Wizard';
import Scene3D from './Scene3D';
import styles from './World.styles';

type Props = {
  containerEl: HTMLDivElement,
  containerWidth: number,
  containerHeight: number,
};

export const World =
({ containerEl, containerWidth, containerHeight }: Props) => (
  <Switch>
    <Route exact path="/editor" component={Wizard} />
    <Route path="/editor/scene3D" render={() => <Scene3D style={styles.container} container={containerEl} width={containerWidth} height={containerHeight} />} />
  </Switch>
);

export default withResizableContainer(World);
