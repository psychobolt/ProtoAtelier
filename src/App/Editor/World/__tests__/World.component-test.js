import React from 'react';
import { shallow } from 'enzyme';

import { World } from '../World.component';

describe('component <World />', () => {
  it('should render correctly with Box by default', () => {
    const wrapper = shallow(<World />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with Sphere', () => {
    const props = {
      location: {
        pathname: '/sphere',
      },
    };
    const wrapper = shallow(<World {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
