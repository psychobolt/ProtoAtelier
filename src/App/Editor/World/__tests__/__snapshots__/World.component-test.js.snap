// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <World /> should render correctly with Box by default 1`] = `
<Switch>
  <Route
    component={[Function]}
    exact={true}
    path="/editor"
  />
  <Route
    path="/editor/scene3D"
    render={[Function]}
  />
</Switch>
`;

exports[`component <World /> should render correctly with Sphere 1`] = `
<Switch>
  <Route
    component={[Function]}
    exact={true}
    path="/editor"
  />
  <Route
    path="/editor/scene3D"
    render={[Function]}
  />
</Switch>
`;
